infra/Pos/Diffusion/Subscription/Dns.hs

```diff
@@ -1,170 +1,172 @@
-        -- Try to subscribe to some peer.
-        -- If they all fail, wait a while before trying again.
-        subscribeToOne subDuration dnsPeersList
-
-    logIOException :: IOException -> IO ()
-    logIOException ioException =
-        traceWith logTrace (Error, sformat ("dnsSubscriptionWorker: "%shown) ioException)
-
-    logDNSError :: DNS.DNSError -> IO ()
-    logDNSError dnsError =
-        traceWith logTrace (Error, sformat ("dnsSubscriptionWorker: "%shown) dnsError)
-
-    -- Find peers via DNS, preserving order.
-    -- In case multiple addresses are returned for one name, they're flattened
-    -- and we forget the boundaries, but all of the addresses for a given name
-    -- are adjacent.
-    findDnsPeers :: Int -> Alts (NodeAddr DNS.Domain) -> IO (Alts NodeId)
-    findDnsPeers index alts = do
-        mNodeIds <- resolveDnsDomains defaultPort alts
-        let (errs, nids_) = partitionEithers mNodeIds
-            nids = mconcat nids_
-        when (null nids)       $ traceWith logTrace (Error, msgNoRelays index)
-        when (not (null errs)) $ traceWith logTrace (Error, msgDnsFailure index errs)
-        return nids
-
-    -- How long to wait before retrying in case no alternative can be
-    -- subscribed to.
-    --
-    -- The formula to calculate the retry interval is
-    -- @
-    -- retryInterval = floor (slotDuration / 2 ^ (number of slots that passed during the longest subscription duration))
-    -- @
-    -- For a @20s@ slot: the values will be @20s@, @10s@, @5s@, @2s@,
-    -- @1s@, @0s@, i.e. if there was a subscription that lasted more than
-    -- @5@ slots we will try to re-subscribe immediately.
-    retryInterval :: Millisecond -> IO Millisecond
-    retryInterval d = do
-        slotDur <- slotDuration
-        let -- slot duration in microseconds
-            slotDurF :: Float
-            slotDurF = fromIntegral $ toMicroseconds slotDur
-            -- Number of slots that passed in the longest subscription duration
-            slotsF :: Float
-            slotsF = (fromIntegral $ toMicroseconds d) / slotDurF
-        return $ fromMicroseconds $ floor $ slotDurF / (2 ** slotsF)
-
-    msgDnsFailure :: Int -> [DNS.DNSError] -> Text
-    msgDnsFailure = sformat ("dnsSubscriptionWorker: DNS failure for index "%int%": "%shown)
-
-    msgNoRelays :: Int -> Text
-    msgNoRelays = sformat ("dnsSubscriptionWorker: no relays found for index "%int)
-
-{-# ANN dnsSubscriptionWorker ("HLint: ignore Use unless" :: String) #-}
+    subscribeTo :: MVar (Map (NodeType, NodeId) Int) -> SubscribeTo IO NodeId
+    subscribeTo peersVar nodeId = do
+        duration <- networkSubscribeTo'
+            logTrace
+            oq
+            BucketBehindNatWorker
+            NodeRelay
+            peersVar
+            keepalive
+            subStatus
+            sendActions
+            nodeId
+        timeToWait <- retryInterval duration <$> slotDuration
+        threadDelay (fromIntegral timeToWait * 1000)
```

Comment:
Let me just note that there is a difference here; previously we computed the max of duration time for all alternative peers and we put `threadDelay` after we tried all the alternative peers; now we're using the previous duration time to wait until we subscribe to the next alternative peer.  This way we will punish more a good peer which is among bad ones (i.e. with short connection times): if there is one good peer and `n` bad ones, now we will wait `n` times for a long time, while previously we'd waited once for a short time and spin quickly through the list of bad peers.  There is a downside of this behaviour: lost bandwidth for trying bad peers too often.  But from the other side, I think we expect more often a single bad peer among good ones: and in this case this works more or less the same.